Add procedure & a step in configuring iPXE (#1910)
theforeman/foreman-documentation: guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc

For more information, see https://ipxe.org/appnote/hardware_drivers[supported hardware on ipxe.org] for a list of hardware drivers expected to work with an iPXE-based boot disk.
* The NIC is compatible with iPXE.

To prepare iPXE environment, you must perform this procedure on all {SmartProxies}.

.Procedure
. Enable the *tftp* and *httpboot* services:
+
[options="nowrap" subs="+quotes,attributes"]
----
# {foreman-installer} --foreman-proxy-httpboot true --foreman-proxy-tftp true
----

. Install the `ipxe-bootimgs` RPM package:
+
[options="nowrap" subs="+quotes,attributes"]
----
# {package-install} ipxe-bootimgs
----

> +

Reviewer: ```suggestion
```

ifdef::foreman-el,katello[]
. On Debian/Ubuntu, install the `ipxe` .deb package:
+
----
# {package-install} ipxe
----

> +

Reviewer: ```suggestion
```

If you want to change the default values in the template, clone the template and edit the clone.

.Procedure

> +

Reviewer: ```suggestion
```

. In the {Project} web UI, navigate to *Hosts* > *Provisioning Templates*, enter `Kickstart default iPXE` and click *Search*.
. Optional: If you want to change the template, click *Clone*, enter a unique name, and click *Submit*.
. Click the name of the template you want to use.
. If you clone the template, you can make changes you require on the *Template* tab.
. Click the *Association* tab and select the operating systems that your host uses.
. Click the *Locations* tab and add the location where the host resides.
. Click the *Organizations* tab and add the organization that the host belongs to.
. Click *Submit* to save the changes.
. Navigate to *Hosts* > *Operating systems* and select the operating system of your host.
. Click the *Templates* tab.
. From the *iPXE Template* list, select the template you want to use.
. Click *Submit* to save the changes.
. Navigate to *Hosts* > *All Hosts*.
. In the *Hosts* page, select the host that you want to use.
. Select the *Operating System* tab.

> . Set *PXE Loader* to *iPXE Embedded*.

Reviewer: ```suggestion
. Set *PXE Loader* to *iPXE Embedded* or *None*.
```
As was suggested in the BZ, perhaps?

Author: Agree :+1:

. From the *iPXE template* list, select the template you want to use.
. Click *Submit* to save the changes.
. Navigate to *Hosts* > *All Hosts*, and select the host you want to use.
. Select the *Operating System* tab.
. Set *PXE Loader* to *PXELinux BIOS* to chainboot iPXE via PXELinux, or to *iPXE Chain BIOS* to load `undionly-ipxe.0` directly.
. Select the *Templates* tab, and from the *PXELinux template* list, select *Review* to verify the template is the correct template.
. From the *iPXE template* list, select *Review* to verify the template is the correct template.
If there is no PXELinux entry, or you cannot find the new template, navigate to *Hosts* > *All Hosts*, and on your host, click *Edit*.
Click the *Operating system* tab and click the Provisioning Template *Resolve* button to refresh the list of templates.
. Configure the `dhcpd.conf` file as follows:
+
[options="nowrap" subs="+quotes,attributes"]
----
if exists user-class and option user-class = "iPXE" {
  filename "http://{smartproxy-example-com}:8000/unattended/iPXE?bootstrap=1";
} # elseif existing statements if non-iPXE environment should be preserved
----
+
[NOTE]

> If you have changed the port using the `--foreman-proxy-http-port installer` option, use your custom port.

Reviewer: ```suggestion
If you have changed the port using the `--foreman-proxy-http-port` installer option, use your custom port.
```

----
+
If you use an isolated network, use a {SmartProxyServer} URL with TCP port `8000`, instead of the URL of {ProjectServer}.
. In creating a system entry to build it using iPXE workflow, ensure selecting the following as the PXELoader for BIOS and UEFI-based builds in the *Operating System* tab of the Host.
+
For BIOS Based systems, The PXELoader can be selected as *None* or *iPXE Embedded*.
+
For UEFI Based systems, The PXELoader must be selected as *None*.

> +

Author: Can we keep this as it is? For better readability.

> +

Reviewer 2: I have no strong objections, so sure.

> iPXE keeps looping while configuring it to work with the Foreman discovery method for UEFI systems.
> To avoid it, follow the below procedure:

Reviewer 2: This looks either like a separate step or like an *Important* admonition to me.

```suggestion
iPXE keeps looping while configuring it to work with the Foreman discovery method for UEFI systems.
To avoid it, enter the following commands:
```

----
+
If you use an isolated network, use a {SmartProxyServer} URL with TCP port `8000`, instead of the URL of {ProjectServer}.
. In creating a system entry to build it using iPXE workflow,
ensure selecting the following as the PXELoader for BIOS and UEFI-based builds in the *Operating System* tab of the host:
+
For BIOS Based systems, PXELoader can be selected as *None* or *iPXE Embedded*.
+
For UEFI Based systems, PXELoader must be selected as *None*.
+
iPXE keeps looping while configuring it to work with the Foreman discovery method for UEFI systems.
To avoid it, enter the following commands:
+
[options="nowrap" subs="+quotes,attributes"]
----
# cp /usr/share/ipxe/ipxe.lkrn /var/lib/tftpboot/
# cp /usr/share/ipxe/undionly.kpxe /var/lib/tftpboot/undionly-ipxe.0
# restorecon -RvF /var/lib/tftpboot/

> ----

Reviewer: I don't think this is true, what prevents the looping is the PXELoader option, the way I understand the BZ.
Additionally, you're covering these commands at the top.

Author: Right. The cherry-picking commit has covered PXELoader and didn't repeat the step. :+1:

+
[NOTE]

> For `\http://{foreman-example-com}/unattended/iPXE`, you can also use a {ProjectName} {SmartProxy} `\http://{smartproxy-example-com}:8000/unattended/iPXE`.

Reviewer: This note belongs with the previous step.

Author: Got fixed automatically with #1726 cherry-picking. :+1:
